escrow/etcd_storage.go

```diff
@@ -0,0 +1,121 @@
+package escrow
+
+import (
+	"bytes"
+	"encoding/gob"
+
+	"github.com/singnet/snet-daemon/etcddb"
+	"github.com/spf13/viper"
+)
+
+// EtcdStorage storage based on etcd
+type EtcdStorage struct {
+	client *etcddb.EtcdClient
+}
+
+// NewEtcdStorage create an etcd storage from Ciper config
+func NewEtcdStorage(vip *viper.Viper) (storage *EtcdStorage, err error) {
+	client, err := etcddb.NewEtcdClient(vip)
+
+	if err != nil {
+		return
+	}
+
+	storage = &EtcdStorage{client: client}
+	return
+}
+
+// Get gets value from etcd by key
+func (storage *EtcdStorage) Get(key *PaymentChannelKey) (state *PaymentChannelData, ok bool, err error) {
+
+	keyBytes, err := serialize(key)
+	if err != nil {
+		return
+	}
+
+	valueBytes, err := storage.client.Get(keyBytes)
+
+	if err != nil || len(valueBytes) == 0 {
+		return
+	}
+
+	state = &PaymentChannelData{}
+	err = deserialize(valueBytes, state)
+
+	if err != nil {
+		return
+	}
+
+	ok = true
+	return
+}
+
+// Put puts key and value to etcd
+func (storage *EtcdStorage) Put(key *PaymentChannelKey, channel *PaymentChannelData) (err error) {
+
+	keyBytes, err := serialize(key)
+	if err != nil {
+		return
+	}
+
+	valueBytes, err := serialize(channel)
+	if err != nil {
+		return
+	}
+
+	err = storage.client.Put(keyBytes, valueBytes)
+	return
+}
+
+// CompareAndSwap uses CAS operation to set a value
+func (storage *EtcdStorage) CompareAndSwap(
+	key *PaymentChannelKey,
+	prevState *PaymentChannelData,
+	newState *PaymentChannelData,
+) (ok bool, err error) {
+
+	keyBytes, err := serialize(key)
+	if err != nil {
+		return
+	}
+
+	expectBytes, err := serialize(prevState)
```

Comment:
As special case previous state can be nil to designate that value is absent in storage. Not sure this code handles situation correctly. Actually I am thinking about adding PutIfAbsent method instead of CAS(key, nil, state). What do you think?

Reply:
I filled an issue on it: https://github.com/singnet/snet-daemon/issues/68